src/mpid/common/shm/mpidu_init_shm.c

```diff
@@ -75,6 +77,51 @@ static int Init_shm_barrier()
     return mpi_errno;
 }
 
+/* FIXME: this is not a scalable algorithm because everyone is polling on the same cacheline */
+static int Init_shm_allsum(int val, int *ret)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_INIT_SHM_ALLSUM);
+
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_INIT_SHM_ALLSUM);
+
+    if (local_size == 1)
+        goto fn_exit;
+
+    MPIR_ERR_CHKINTERNAL(!barrier_init, mpi_errno, "allsum not initialized");
+
+    /* First add up the values before anyone queries the results. */
+    OPA_add_int(&barrier->val, val);
+    if (OPA_fetch_and_incr_int(&barrier->participants) == local_size - 1) {
+        OPA_store_int(&barrier->wait, 1 - sense);
+        OPA_write_barrier();
+    } else {
+        /* wait */
+        while (OPA_load_int(&barrier->wait) == sense)
+            MPL_sched_yield();  /* skip */
+    }
+    sense = 1 - sense;
+
```

Comment:
Can we reuse `Init_shm_barrier` here ?

Reply:
Besides, it looks like the use of `barrier->val` is not consistent. It is used to count the arrived processes in `Init_shm_barrier`, but used to calculate the sum here.  I also feel confused that the `barrier` struct is used for `allsum`. Maybe we can define the internal structure as:
```
typedef struct Init_shm_barrier {
    OPA_int_t participants;
    OPA_int_t wait;
} Init_shm_barrier_t;

typedef struct Init_shm_reduce {
    OPA_int_t participants;
    OPA_int_t wait;
    OPA_int_t val;
} Init_shm_reduce_t;
	
typedef struct Init_shm_coll {
    union {
      Init_shm_barrier_t barrier;
      Init_shm_reduce_t reduce;
    }
} Init_shm_coll_t;

static Init_shm_coll_t *internal_coll;
``` 
Alternatively, we can use the existing `Init_shm_put|get|barrier` routines to implement allsum in XPMEM. The `Init_shm` utility is used only at MPI init, thus performance is not a big concern.

@gcongiu Can you please also check the `Init_shm` related change ? I am not very sure if you plan to change it and if the extension aligns with the final design.

Reply:
Hmm. I do like that `union` scheme. That's much cleaner without increasing the footprint. I'll make that change.

As for switching to just using`put`/`get`/`barrier`, you might be close to right. Now that I think of it, we can probably just implement this using a compare and swap and a barrier. 🤔 

That's probably the better way to go with minimal change. What do you think?

Reply:
Arg...I didn't realize there wasn't a `MPIDU_Init_shm_cas` function already. So that's not actually any easier. 

As for implementing it with puts and gets... I'm not sure what you're proposing unless you mean everyone writes in order and barriers in between.

Reply:
How about this approach ? It does not extend the `MPIDU_Init_shm` API.
```
MPIDI_XPMEM_mpi_init_hook(...) {
    ...
    MPIDU_Init_shm_put(&MPIDI_XPMEM_global.segid, sizeof(xpmem_segid_t));
    MPIDU_Init_shm_barrier();
    
    bool anyfail = false;
    for (i = 0; i < num_local; i++) {
        MPIDU_Init_shm_get(i, sizeof(xpmem_segid_t), &MPIDI_XPMEM_global.segmaps[i].remote_segid);
        if(MPIDI_XPMEM_global.segmaps[i].remote_segid == -1) {
            anyfail = true;
            break;
        }
    }
    if(anyfail) {
        /* disable xpmem and release local segid */
       goto fn_exit;
    }

    /* Initialize segmap for every local processes */
}
```

Reply:
👍 I like that. It's mostly the same as what's there already.

Reply:
Ok. I've pushed the change for this.